src/util/metadata/side_metadata/global.rs

```diff
@@ -97,6 +99,515 @@ impl SideMetadataSpec {
+    pub fn fetch_add_atomic<T: MetadataValue>(
+        &self,
+        data_addr: Address,
+        val: T,
+        order: Ordering,
+    ) -> T {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                let bits_num_log = self.log_num_of_bits;
+                if bits_num_log < 3 {
+                    FromPrimitive::from_u8(self.fetch_ops_on_bits(
+                        data_addr,
+                        meta_addr,
+                        order,
+                        order,
+                        |x: u8| x.wrapping_add(val.to_u8().unwrap()),
+                    ))
+                    .unwrap()
+                } else {
+                    unsafe { T::fetch_add(meta_addr, val, order) }
+                }
+            },
+            |_old_val| {
+                #[cfg(feature = "extreme_assertions")]
+                // sanity::typed_verify_add(self, data_addr, val, _old_val)
+                sanity::verify_update::<T>(self, data_addr, _old_val, _old_val.wrapping_add(&val))
+            },
+        )
+    }
+
+    #[inline(always)]
+    pub fn fetch_sub_atomic<T: MetadataValue>(
+        &self,
+        data_addr: Address,
+        val: T,
+        order: Ordering,
+    ) -> T {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                if self.log_num_of_bits < 3 {
+                    FromPrimitive::from_u8(self.fetch_ops_on_bits(
+                        data_addr,
+                        meta_addr,
+                        order,
+                        order,
+                        |x: u8| x.wrapping_sub(val.to_u8().unwrap()),
+                    ))
+                    .unwrap()
+                } else {
+                    unsafe { T::fetch_sub(meta_addr, val, order) }
+                }
+            },
+            |_old_val| {
+                #[cfg(feature = "extreme_assertions")]
+                // sanity::typed_verify_sub(self, data_addr, val, _old_val)
+                sanity::verify_update::<T>(self, data_addr, _old_val, _old_val.wrapping_sub(&val))
+            },
+        )
+    }
+
+    #[inline(always)]
+    pub fn fetch_and_atomic<T: MetadataValue>(
+        &self,
+        data_addr: Address,
+        val: T,
+        order: Ordering,
+    ) -> T {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                if self.log_num_of_bits < 3 {
+                    FromPrimitive::from_u8(self.fetch_ops_on_bits(
+                        data_addr,
+                        meta_addr,
+                        order,
+                        order,
+                        |x: u8| x & val.to_u8().unwrap(),
+                    ))
```

Comment:
`fetch_and` doesn't need cmpxchg.  Just set irrelevant bits to 1 and they will not be affected.

```suggestion
let mask = meta_byte_mask(self) << lshift;
let opnd = val | (-1 & !mask);
unsafe { T::fetch_and(meta_addr, opnd, order) };
```

Reply:
Sounds good. In case you may be interested, the following is the assembly code of `fetch_and()` for 1 bit side metadata when using `fetch_ops_on_bits` (internally uses `fetch_update`), and directly using `fetch_and`:

### with fetch_ops_on_bits
```asm
mmtk::util::alloc_bit::unset_alloc_bit_unsafe:
 mov     rcx, rdi
 shr     rdi, 6
 shr     cl, 3
 movabs  rsi, 13194139533312
 mov     al, byte, ptr, [rdi, +, rsi]
 mov     dl, -2
 rol     dl, cl
.LBB225_1:
 mov     ecx, eax
 and     cl, dl
 lock    cmpxchg, byte, ptr, [rdi, +, rsi], cl
 jne     .LBB225_1
 ret
```
### with fetch_and
```asm
mmtk::util::alloc_bit::unset_alloc_bit_unsafe:
 mov     rcx, rdi
 mov     rax, rdi
 shr     rax, 6
 shr     cl, 3
 mov     dl, -2
 rol     dl, cl
 movabs  rcx, 13194139533312
 lock    and, byte, ptr, [rax, +, rcx], dl
```